File: day-03/solution.ts
import { readLines } from "../commons/io.ts";

function main() {
  const lines = readLines("day-03/input.txt");
  const ALPHABET = "_abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";

  let prioritySum1 = 0;

  // PART 01

  lines.forEach((line) => {
    // split line into 2 parts at mid point
    const first = line.slice(0, line.length / 2);
    const second = line.slice(line.length / 2);
    // find elements that are in both first and second
    const intersection = first
      .split("")
      .filter((char) => second.includes(char));
    const priority = ALPHABET.indexOf(intersection[0]);
    prioritySum1 += priority;
  });

  // PART 02
  let prioritySum2 = 0;
  // group lines into groups of 3
  for (let i = 0; i < lines.length; i += 3) {
    const first = lines[i];
    const second = lines[i + 1];
    const third = lines[i + 2];
    // find elements that are in all 3 lines
    const intersection = first
      .split("")
      .filter((char) => second.includes(char) && third.includes(char));
    const priority = ALPHABET.indexOf(intersection[0]);
    prioritySum2 += priority;
  }

  console.log("Part 1", prioritySum1);
  console.log("Part 2", prioritySum2);
}

main();

File: day-03/solution.py
import sys
import os
from typing import List

sys.path.append(os.path.join(os.path.dirname(__file__), ".."))

from commons.io import readLines


def main():
  lines: List[str] = readLines("day-03/input.txt")
  ALPHABET: str = "_abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"

  prioritySum1: int = 0

  # PART 01

  for line in lines:
    # split line into 2 parts at mid point
    first = line[0:len(line) // 2]
    second = line[len(line) // 2:]
    # find elements that are in both first and second
    intersection = list(set(first) & set(second))
    priority = ALPHABET.index(intersection[0])
    prioritySum1 += priority

  # PART 02
  prioritySum2 = 0
  # group lines into groups of 3
  for i in range(0, len(lines), 3):
    first = lines[i]
    second = lines[i + 1]
    third = lines[i + 2]
    # find elements that are in all 3 lines
    intersection = [char for char in first if char in second and char in third]
    priority = ALPHABET.index(intersection[0])
    prioritySum2 += priority

  print("Part 01: " + str(prioritySum1))
  print("Part 02: " + str(prioritySum2))


main()
